[css-text-4] Add word-boundary-detection and word-boundary-expansion

frivoal opened the pull request.

fantasai commented on css-text-4/Overview.bs:
```diff
@@ -31,8 +35,519 @@ Introduction</h2>
+	A <dfn data-dfn-for='' data-dfn-type=dfn>virtual word boundary</dfn> is similar to the presence
+	of the ZERO WIDTH SPACE (U+200B) character:
+	it introduces a [=soft wrap opportunity=]
+	and is affected by the 'word-boundary-expansion' property;
```
End sentence.

fantasai commented on css-text-4/Overview.bs:
```diff
@@ -31,8 +35,519 @@ Introduction</h2>
+	and is affected by the 'word-boundary-expansion' property;
+	its presence has no effect on text shaping,
+	nor on 'word-spacing'.
+	However, its insertion must have no effect on the underlying content,
```
Inserting [=virtual word boundaries=] must have ...

fantasai commented on css-text-4/Overview.bs:
```diff
@@ -31,8 +35,519 @@ Introduction</h2>
+	<dl dfn-type=value dfn-for=word-boundary-detection>
+		<dt><dfn>manual</dfn>
+		<dd>
+			The User Agent must not insert [=virtual word boundaries=].
```
Add some more explanatory text for authors, maybe:

Linguistic analysis is not used to determine [=line wrapping opportunities=] in any language or writing system.

or something better...

fantasai commented on css-text-4/Overview.bs:
```diff
@@ -31,8 +35,519 @@ Introduction</h2>
+			with both the [=content language=] and <<lang>>
+			<i>canonicalized</i> and converted to <i>extlang form</i>
+			as per section 4.5 of [[!RFC5646]],
+			then the [=used value=] is ''word-boundary-detection/manual'',
```
Should this be normal?

fantasai commented on css-text-4/Overview.bs:
```diff
@@ -31,8 +35,519 @@ Introduction</h2>
+			(see [[BCP47]]).
+			If the UA does not support word-boundary detection
+			for <em>all</em> languages represented by the specified range,
+			it must reject that value at parse-time.
```
"rejected at parse-time" is not the right terminology. You want to say it's invalid. E.g.

If the UA does not support word-boundary detection for all languages represented by the specified language range, that specified value is invalid (and will cause the declaration to be ignored).

or

Only `<<lang>>` values for which the UA has support of all languages represented by the specified range are valid.

or somesuch.

fantasai commented on css-text-4/Overview.bs:
```diff
@@ -31,8 +35,519 @@ Introduction</h2>
+	</div>
+
+	<div class=example>
+		Specifying the language for which the word boundary detection is to be performed
```
and making unsupported language ranges invalid

fantasai commented on css-text-4/Overview.bs:
```diff
@@ -31,8 +35,519 @@ Introduction</h2>
+			between a [=typographic letter unit=]
+			and a subsequent [=typographic character unit=] from the [[!UNICODE]] Pe or Pf classes,
+			or between a [=typographic letter unit=]
+			and a preceeding [=typographic character unit=] from the [[!UNICODE]] Ps or Pi classes,
```
IIRC Pf/Pi classes are characters that are sometimes used as opening and sometimes used as closing, so this might not be quite correct.

frivoal replied:
Good point. I've though some more about it, and derived updated rules from UAX14 line breaking classes, rather than directly from unicode general categories.

fantasai commented on css-text-4/Overview.bs:
```diff
@@ -31,8 +35,519 @@ Introduction</h2>
+	</dl>
+
+	The User Agent must not replace
+	instances of U+200B imediately preceding or following
```
immediately

fantasai commented on css-text-4/Overview.bs:
```diff
@@ -31,8 +35,519 @@ Introduction</h2>
+	However, it is defined as a separate property
+	rather than additional values to 'text-transform' because:
+	* This property needs to take effect before [[CSS-TEXT-3#white-space-rules]],
+		but 'text-transform' happens after that.
```
explain why

frivoal commented:
Ok, I've addressed your comments and merged. The largest change I made was in response to your comment about Pi an Pf. Thinking about it, not only did I agree, but I though this section would work out better using UAX14 line breaking classes than unicode general categories. See https://github.com/w3c/csswg-drafts/pull/4153/commits/87a41925005e2f67a3738f5bc840a884d369d754

If you think that's actually a step back, let me know (here or in a new issue), and I'll revert to what it was prior to that, with Pi and Pf removed.
